### PR TITLE
Feature/process manager

### LINKED ISSUE #9: Implement Process Manager
The process manager will be getting jobs request from the kafka stream/Global Comm Manager and distribute the tasks amongst the already running child processes.

The data scientist will specify the `number of workers` in the process pool in the beginning and will launch always running child processes.

Each child process holds a `BaseTrainer` object and a job queue. For every training job, we call `laod_state` with [serialized worker state](https://github.com/NimbleEdge/RecoEdge/blob/main/fedrec/federated_worker.py#L35)

```python
trainer.load_state(
            state.state_dict['model'],
            state.state_dict['optimizer'],
            state.model_preproc)
```

Once the job is finished the response goes to the global com stream which then unsuspends the worker routine waiting for the response.

### PR DESCRIPTION
**Pull request by harshv834**
Fixes #9

### CODE REVIEW

File: fedrec/multiprocessing/process_manager.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+                job_request = self.process_comm_manager.receive_message()
+                if job_request is not None:
+                    if job_request["job_type"] == "END":
+                        self.loop.stop()
+                    self.enqueued_jobs.append(self,job_request)
+
+
+    async def run_jobs(self) -> None:
+        while True:
+            if len(self.enqueued_jobs) > 0:
```
Comment (harshv834) [inline fedrec/multiprocessing/process_manager.py:37]: Make enqueued_jobs into asyncio queue.

File: fedrec/multiprocessing/process_manager.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        self.loop.create_task(self.run_jobs())
+        self.loop.run_forever()
+
+    async def consume(self) -> None:
+        while True:
+            if len(self.enqueued_jobs) < self.max_jobs_per_process:
+                job_request = self.process_comm_manager.receive_message()
+                if job_request is not None:
+                    if job_request["job_type"] == "END":
+                        self.loop.stop()
```
Comment (harshv834) [inline fedrec/multiprocessing/process_manager.py:31]: Add all cancellation types.

File: fedrec/multiprocessing/process_manager.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import mpi4py
+from fedrec.communications.process_com_manager import ProcessComManager
+from fedrec.utilities import registry
+from fedrec.multiprocessing.job import Jobber
+from mpi4py import MPI
+import asyncio
+class MPIProcessManager:
+
+    def __init__(self, config) -> None:
+        self.pool = MPI.COMM_WORLD
```
Comment (harshv834) [inline fedrec/multiprocessing/process_manager.py:10]: main process doesn't run process manager.

File: fedrec/communications/process_com_manager.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    def run(self):
+        super().run()
+
+    @tag_reciever(JobMessage.JOB_REQUEST)
+    def send_message(self, msg_params):
+        #TODO : Send job completion message to worker comm manager.
+        raise NotImplementedError
+
+    @tag_reciever(JobMessage.JOB_COMPLETION)
+    async def receive_message(self,):
```
Comment (harshv834) [inline fedrec/communications/process_com_manager.py:20]: Tasks  ---
- Handle message like worker com manager
- Process-related messages write here.

File: fedrec/multiprocessing/process_manager.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+                job_request = self.enqueued_jobs.pop(0)
+                job = self.loop.create_task(self.jobber.run(job_request))
+                job.add_done_callback(self.publish())
+
+
+
+    def publish(self, job_result) -> None:
+        self.process_comm_manager.send_message(job_result.result())
+
+    def balance_load(self):
```
Comment (harshv834) [inline fedrec/multiprocessing/process_manager.py:47]: Remove this from here.

File: fedrec/multiprocessing/job.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+class Jobber:
+    """
+    Jobber class only handles job requests based on job type
+    """
+    def __init__(self, trainer, logger) -> None:
+        self.logger = logger
+        self.trainer: BaseTrainer = trainer
+
+    def run(self, message_dict):
+        print("here")
```
Comment (harshv834) [inline fedrec/multiprocessing/job.py:13]: Format job type input and output according to PR #13

File: fedrec/communications/process_com_manager.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
+
+from fedrec.utilities import registry
+from fedrec.communications.messages import JobMessage
+from fedrec.communications.comm_manager import (CommunicationManager,
+                                                tag_reciever)
+
+
+@registry.load("process_comm_manager", "zeroMQ_process_comm_manager")
```
Comment (vkkhare) [inline fedrec/communications/process_com_manager.py:8]: a shorter name will help ;p

Comment (vkkhare) [reply fedrec/communications/process_com_manager.py:8]: Btw why do we need a registry here? It's not doing anything specific to zeroMQ

**Review by vkkhare — CHANGES_REQUESTED** [review]

File: fedrec/multiprocessing/process_manager.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+class MPIProcessManager:
+
+    def __init__(self, config, trainer, logger) -> None:
+        self.pool = MPI.COMM_WORLD
+        self.rank = self.pool.Get_rank()
+        self.num_processes = self.pool.Get_size()
+        if self.rank !=0:
+            self.jobber = Jobber(trainer = trainer, logger = logger)
+            self.enqueued_jobs = asyncio.Queue(maxsize=config["max_jobs_per_process"])
+            self.process_comm_manager = registry.construct("process_comm_manager", config_dict = config["comm_manager_config"])
```
Comment (vkkhare) [inline fedrec/multiprocessing/process_manager.py:17]: Will process_com_manager be different from communication interfaces like zmq? If not we can probably subclass this to communication manager

File: fedrec/multiprocessing/process_manager.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+
+
+    def run(self) -> None:
+        if self.rank != 0:
+            self.loop.create_task(self.consume())
+            self.loop.create_task(self.run_jobs())
+            self.loop.run_forever()
+
+    async def consume(self) -> None:
```
Comment (vkkhare) [inline fedrec/multiprocessing/process_manager.py:28]: If we subclass this to communication manager, the consume could be independent or overriden accordingly

**Review by vkkhare — APPROVED** [review]

**Pull request merged**